### DISCUSSION TITLE
introduce track type objects

### DISCUSSION THREAD

**Discussion by karst** [OpenRCT2/OpenRCT2#17892] [Feature requests]
I had been thinking what might be the smartest way to approach ride type objects and I had the following suggestion about that.

Upgrade ride objects
-----
Instead of ride type objects you can make a custom ride type to be part of a vehicle. To do this you will require to choose a track type ID in the json to what track it renders as by default. This means a ride object can be completely custom and not just a vehicle on an existing ride type. To get it to function correctly you would have to fill in all details currently found in the meta files of the rides.

Introduce track type objects
-----
What would a track type object be? This is the track that gets rendered, similar to how it's currently done with the B&M coaster types. This can be used by ride types that include its ID in their json file as `defaultTrackType` (for example `"defaultTrackType": "rct2.track.bm"`). In this file we need to also have listed what track elements are available to it and if it includes alternative textures . Ride objects should be able to choose what alternative textures should be used. For example the steel looping coaster could have a track version with a chainlift and a track version with tyre drives, but is still the same track type in other aspects. Or think of for example the looping which is different between steel looping and LIM while they use the same track type, or even entirely swappable pieces like discussed in #17583.

I don't know if this is the smartest way to implement this functionality but it should be pretty helpful. I think it should make things overall easier with not having duplicate code within openrct2 itself. It would be a breeze to replace/swap textures with asset packs too I imagine. And ofcourse it would make it easier to mod the game like @X123M3-256 is doing with his custom tracks right now. (Meaning people wont have to use a separate branch to use that). 

As a sidenote I think it would be appreciated if the track renderer, ride generator and such would be officially supported for content creators to use that know how to model but not know how to program.

If there's more ideas on how to implement this leave it in the discussion below.

**Comment by frutiemax** [comment]
It's planned to transform the ride types into objects, that's why you see all the PRs that remove the RIDE_TYPE_* references.

**Reply by karst** [reply]
Yea, i just think that upgrading the existing ride objects and introducing track type objects would cause less clutter/double code, size wise and optimization wise than introducing ride type objects. Functionality and stats can be done by the ride object. Rendering of the track is done by track type objects.

**Comment by karst** [comment]
I think for this implementation important would be:

Parent object: allowing linking the object to another object, if it has a few different pieces, for example brakes, from its parent object, but otherwise have the same sprites.
paint code under track style:
default is track style 0
allows for multiple track styles, would require their own icons when multiple are provided (example is the dingy slide)
settings for track style behaviour metadata.
Setting custom bounding boxes (optional, will use default if not provided).

will update if i think of more

**Reply by X123M3-256** [reply]
> allows for multiple track styles, would require their own icons when multiple are provided (example is the dingy slide)

I think each track style should be a separate object in cases like this. Especially as many rides might share track styles with another ride but not *all* styles (for example, the multi looping coaster ought to have the two and three rail track but I wouldn't give it four rail by default).

**Comment by Realsteel89** [comment]
the ride generator must surely only be rewritten like you did it with the track generator. It has lines of code that windows cannot interpret correctly, e.g. "paths". I'm currently trying to get the ride render to run think that should also run soon under windows. in addition, I wanted to incorporate this then in my Ui.

**Comment by spacek531** [comment]
The wooden roller coaster has gotten some paint code added that takes advantage of the `SpriteBoundBox2` struct that @Gymnasiast introduced. A first step in object-defined track paint could be to make a template function that uses the track's type to reference a struct that contains all the information necessary to draw the piece, which the template function then draws. This could be first implemented on wooden coaster (since it has some track pieces using SpriteBoundBox2 already) or junior coaster (since it is the classical testbed for paint code changes). Having a template function allows individual track pieces to be converted without requiring all of the track pieces converted at once, using the current implementation of paint code function getter - the getter would simply return the same function for multiple pieces differentiated only by the template argument. When all track pieces are implemented, the getter can simply return a single function with the template argument, and when all track types are implemented, the getter function can be dropped in favor of just calling the one function that paints the track piece.
